sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+      findClpAnnotation(clazz) match {
+        case None => throw new BadAnnotationException(s"The class '${clazz.getSimpleName}' is missing the required @clp annotation.")
+        case Some(clp) =>
+          clpsToProperty.put(clazz, clp)
+          val clpGroup: ClpGroup = clpGroupClassToClpGroupInstance.get(clp.group) match {
+            case Some(group) => group
+            case None =>
+              val group: ClpGroup = clp.group.newInstance
+              clpGroupClassToClpGroupInstance.put(clp.group, group)
+              group
+          }
+          Option(clpsByGroup.get(clpGroup)) match {
+            case Some(clps) =>
+              clps.add(clazz)
+            case None =>
+              val clps = ListBuffer[ClpClass](clazz)
+              clpsByGroup.put(clpGroup, clps)
+          }
+      }
+    }
+
+    clpsByGroup.entrySet.foreach{entry =>
+      val clpGroup: ClpGroup = entry.getKey
+      builder.append(SeparatorLine)
+      builder.append(wrapString(KRED, String.format(s"%-${ClpGroupNameColumnLength}s%-${ClpGroupDescriptionColumnLength}s\n",
+        clpGroup.name + ":", clpGroup.description), KNRM))
+      val entries: List[ClpClass] = entry.getValue.toList
+      entries
+        .sortWith((lhs, rhs) => lhs.getSimpleName.compareTo(rhs.getSimpleName) < 0)
+        .foreach{clazz =>
+          val clp: clp = clpsToProperty.get(clazz).get
+          builder.append(wrapString(KGRN, String.format(s"    %-${ClpNameColumnLength}s%s\n",
+            clazz.getSimpleName, wrapString(KCYN, formatShortDescription(clp.description))), KNRM))
+        }
+    }
+
+    builder.append(SeparatorLine)
+    builder.toString()
+  }
+
+  /** The main method for parsing command line arguments.  Typically, the arguments are the format
+    * `[clp Name] [clp arguments]`.  This in contrast to [[parseMainAndClp]], which includes arguments to the main class in
+    * the `[main arguments]` section.
+    *
+    *   (1) Finds all classes that extend [[Clp]] to include them as valid command line programs on the command line.
+    *   (2) When the first arg is one of the command line programs names (usually `getSimpleName`), it creates a
+    *       [[CommandLineProgramParser]] (Clp parser) for the given command line program, and allows the Clp parser to
+    *       parse the rest of the args.
+    *   (3) Returns `None` in the case of any parsing error, or an instance of the command line program constructed
+    *       according to the parsed arguments.
+    *
+    * @param args              the command line arguments to parse.
+    * @param packageList       the list of packages to search for classes that extend [[Clp]].
+    * @param omitSubClassesOf  individual classes to omit from including on the command line.
+    * @param includeHidden     true to include classes whose [[clp]] annotation's hidden values is set to false.
+    * @param clpBlock          a block of code to execute after a [[clp]] }instance has been created successfully.
```

Comment:
Imma gonna win this argument :stuck_out_tongue:  `clpBlock` here is actually a function type because it's `Clp => Unit` - i.e. it has a parameter list and a return type.  `block` in `yieldAndThen` is a call-by-name parameter, which is (frustratingly) not a function type in scala, hence the need to play silly tricks like this in EitherTask:
`  def apply(left: Task, right: Task, choice: => Choice): EitherTask = new EitherTask(left, right, () => choice)`

Ahh, technically right, the best kind of right :wink: